src/test/clone/test_fork.rs

```diff
@@ -116,6 +116,110 @@ fn test_clone_parent(set_clone_parent: bool) -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
+fn test_child_change_session() -> Result<(), Box<dyn Error>> {
+    let (reader, writer) = rustix::pipe::pipe().unwrap();
+
+    let parent_sid = unsafe { libc::getsid(0) };
+    let parent_pgrp = unsafe { libc::getpgrp() };
+
+    let clone_res = unsafe { linux_api::sched::fork() }.unwrap();
+    let child_pid = match clone_res {
+        CloneResult::CallerIsChild => {
+            // Ensure we exit with non-zero exit code on panic.
+            std::panic::set_hook(Box::new(|info| {
+                eprintln!("panic: {info:?}");
+                unsafe { libc::exit(1) };
+            }));
+
+            // Should initially be in same session and group as parent.
+            assert_eq!(unsafe { libc::getpgrp() }, parent_pgrp);
+            assert_eq!(unsafe { libc::getsid(0) }, parent_sid);
+
+            // Change session
+            let pid = unsafe { libc::getpid() };
+            assert_eq!(unsafe { libc::setsid() }, pid);
+
+            // Should now be session and group leader.
+            assert_eq!(unsafe { libc::getpgrp() }, pid);
+            assert_eq!(unsafe { libc::getsid(0) }, pid);
+
+            assert_eq!(rustix::io::write(&writer, &[0]), Ok(1));
+            unsafe { libc::exit(0) };
+        }
+        CloneResult::CallerIsParent(child_pid) => child_pid,
+    };
+
+    // Close our copy of the writer-end of the pipe, so that parent hang trying
+    // to read from the pipe if the child exited abnormally.
+    drop(writer);
```

Comment:
I don't follow this. By closing this process' writer, doesn't this make this process' `read` *not* hang if the child exited abnormally? I'm not sure if just the comment is wrong here. If there are no writers left, then a `read(reader, &buf, 1)` should return 0.